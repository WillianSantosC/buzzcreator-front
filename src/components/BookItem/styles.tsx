import { styled } from "../../../styled-system/jsx";

export const Wrapper = styled("div", {
  base: {
    padding: "token(spacings.small)",
    borderBottom: "0.1rem solid token(colors.lightGray)",
    display: { _mdMI: "flex" },
  },
});

export const AlterWrapper = styled("div", {
  base: {
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    gap: { base: "1rem" },
  },
});

export const Controls = styled("div", {
  base: {
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    gap: { base: "1rem" },
    border: "solid 0.1rem token(colors.darkGray)",
    borderRadius: "token(border.radius)",
    padding: "0.3rem",

    "& button": {
      border: "none",
      outline: "none",
      padding: "0.3rem 0.8rem",
      margin: "0",
      font: "bold",
      fontSize: "token(font.sizes.xsmall)",
      background: "token(colors.lightGray)",
      borderRadius: "token(border.radius)",
    },
  },
});

export const InfoWrapper = styled("div", {
  base: {
    display: "flex",
    alignItems: "center",
    gap: { base: "1rem" },
    width: "20%",
    justifyContent: "space-between",
  },
});

export const Icon = styled("img", {
  base: {
    width: "2rem",
    height: "2rem",
    cursor: "pointer",
  },
});

export const BookContent = styled("div", {
  base: {
    display: "flex",
  },
});

export const ImageBox = styled("div", {
  base: {
    flexShrink: "0",
    marginRight: { base: "1.2rem", _mdMI: "token(spacings.xsmall)" },
    width: { base: "9.6rem", _mdMI: "15rem" },
    height: { base: "5.6rem", _mdMI: "7rem" },

    "& img": {
      width: "100%",
      height: "100%",
      objectFit: "cover",
    },
  },
});

export const Content = styled("div", {
  base: {
    display: "flex",
    flexDirection: "column",
    alignItems: "flex-start",
  },
});

export const Title = styled("h3", {
  base: {
    fontSize: { base: "token(font.sizes.small)", _mdMI: "2rem" },
    lineHeight: { base: "token(font.sizes.small)", _mdMI: "2rem" },
    fontWeight: "token(font.bold)",
    color: "token(colors.black)",
    marginBottom: "token(spacings.xxsmall)",
  },
});

export const Price = styled("div", {
  base: {
    color: "token(colors.white)",
    padding: "0.2rem token(spacings.xxsmall)",
    backgroundColor: "token(colors.secondary)",
    borderRadius: "token(border.radius)",
    fontSize: "token(font.sizes.small)",
    fontWeight: "token(font.bold)",
  },
});

export const DownloadLink = styled("a", {
  base: {
    color: "token(colors.primary)",
    marginLeft: "token(spacings.xxsmall)",
  },
});

export const PaymentContent = styled("div", {
  base: {
    color: "token(colors.gray)",
    fontSize: "token(font.sizes.small)",
    display: "flex",
    minWidth: "28rem",
    flexDirection: { base: "column", _mdMI: "column-reverse" },
    marginTop: { base: "token(spacings.xsmall)", _mdMI: "0" },
    flex: { _mdMI: "1" },
    justifyContent: { _mdMI: "space-between" },
    alignItems: { _mdMI: "flex-end" },
  },
});

export const CardInfo = styled("div", {
  base: {
    display: "flex",
    alignItems: "center",
    marginTop: { _mdMI: "token(spacings.xsmall)" },

    "& img": {
      marginLeft: "token(spacings.xxsmall)",
    },
  },
});
