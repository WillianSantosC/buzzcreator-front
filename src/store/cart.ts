import { create } from "zustand";
import { persist } from "zustand/middleware";

export type CartItem = {
  id: number;
  titulo: string;
  preco: number;
  imagem: string;
  estoque: number;
  quantidade: number;
};

type CartStore = {
  items: CartItem[];
  addItem: (item: CartItem) => void;
  removeItem: (id: number) => void;
  updateQuantity: (id: number, quantity: number) => void;
  clearCart: () => void;
  total: () => number;
};

export const useCartStore = create<CartStore>()(
  persist(
    (set, get) => ({
      items: [],
      addItem: (item) => {
        const existing = get().items.find((i) => i.id === item.id);
        if (existing) {
          set({
            items: get().items.map((i) =>
              i.id === item.id
                ? { ...i, quantidade: i.quantidade + item.quantidade }
                : i,
            ),
          });
        } else {
          set({ items: [...get().items, item] });
        }
      },
      removeItem: (id) => {
        set({ items: get().items.filter((i) => i.id !== id) });
      },
      updateQuantity: (id, quantity) => {
        if (quantity <= 0) {
          get().removeItem(id);
        } else {
          set({
            items: get().items.map((i) =>
              i.id === id ? { ...i, quantidade: quantity } : i,
            ),
          });
        }
      },
      clearCart: () => set({ items: [] }),
      total: () =>
        get().items.reduce(
          (acc, item) => acc + item.preco * item.quantidade,
          0,
        ),
    }),
    {
      name: "cart-storage", // chave do localStorage
    },
  ),
);
