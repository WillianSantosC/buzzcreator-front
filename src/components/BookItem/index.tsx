import * as S from "./styles";

export type BookItemProps = {
  id: number;
  imagem: string;
  titulo: string;
  estoque: number;
  preco: number;
  remove?: boolean;
};

const BookItem = ({ imagem, titulo, preco, remove = false }: BookItemProps) => (
  <S.Wrapper className="BookItemWrapper">
    <S.BookContent>
      <S.ImageBox>
        <img src={imagem} alt={titulo} width={100} height={100} />
      </S.ImageBox>

      <S.Content>
        <S.Title>{titulo}</S.Title>

        <S.InfoWrapper>
          <S.Price>
            {preco.toLocaleString("pt-BR", {
              style: "currency",
              currency: "BRL",
            })}
          </S.Price>

          {remove && <S.Icon src="/img/trash-bl.svg" alt="Delete" />}
        </S.InfoWrapper>
      </S.Content>
    </S.BookContent>
  </S.Wrapper>
);

export default BookItem;
