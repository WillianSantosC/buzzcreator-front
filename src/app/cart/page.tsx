"use client";

import Cart, { CartProps } from "@/templates/Cart";

import { useCartStore } from "@/store/cart";

export default function CartPage() {
  const items = useCartStore((state) => state.items);
  const total = useCartStore((state) => state.total());

  const props: CartProps = {
    items,
    total: total.toLocaleString("pt-BR", {
      style: "currency",
      currency: "BRL",
    }),
  };

  return <Cart {...props} />;
}
