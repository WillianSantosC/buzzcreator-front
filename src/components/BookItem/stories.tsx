import { Meta, StoryObj } from "@storybook/react";

import BookItem, { BookItemProps } from ".";

export default {
  title: "components/BookItem",
  component: BookItem,
  parameters: {
    layout: "fullscreen",
  },
  args: {
    img: "https://www.outerspace.com.br/wp-content/uploads/2018/04/reddeadredemption2.jpg",
    title: "Red Dead Redemption 2",
    price: "R$ 215,00",
  },
} as Meta;

export const Default: StoryObj<BookItemProps> = {};
