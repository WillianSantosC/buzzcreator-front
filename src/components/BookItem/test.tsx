import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import BookItem, { BookItemProps } from ".";

// Mock da store Zustand
jest.mock("../../store/cart", () => ({
  useCartStore: jest.fn(),
}));

import { useCartStore } from "@/store/cart";

const mockUpdateQuantity = jest.fn();
const mockRemoveItem = jest.fn();

const props: BookItemProps = {
  id: 1,
  imagem: "/img/books.jpg",
  titulo: "The Pragmatic Programmer",
  estoque: 5,
  preco: 120,
  remove: true,
};

describe("<BookItem />", () => {
  beforeEach(() => {
    jest.clearAllMocks();

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (useCartStore as unknown as jest.Mock).mockImplementation((selector: any) =>
      selector({
        items: [{ id: props.id, quantidade: 2 }],
        updateQuantity: mockUpdateQuantity,
        removeItem: mockRemoveItem,
      }),
    );
  });

  it("should render the book image with correct alt text", () => {
    render(<BookItem {...props} />);
    const image = screen.getByRole("img", {
      name: /the pragmatic programmer/i,
    });
    expect(image).toHaveAttribute("src", props.imagem);
  });

  it("should render the book title", () => {
    render(<BookItem {...props} />);
    expect(screen.getByText(/the pragmatic programmer/i)).toBeInTheDocument();
  });

  it("should render quantity and increment/decrement buttons", () => {
    render(<BookItem {...props} />);

    expect(screen.getByText("2")).toBeInTheDocument(); // quantidade mockada

    const incrementButton = screen.getByText("+");
    const decrementButton = screen.getByText("-");

    expect(incrementButton).toBeInTheDocument();
    expect(decrementButton).toBeInTheDocument();
  });

  it("should call updateQuantity when increment button is clicked", () => {
    render(<BookItem {...props} />);
    const incrementButton = screen.getByText("+");
    fireEvent.click(incrementButton);
    expect(mockUpdateQuantity).toHaveBeenCalledWith(props.id, 3); // quantidade mockada: 2 + 1
  });

  it("should call updateQuantity when decrement button is clicked", () => {
    render(<BookItem {...props} />);
    const decrementButton = screen.getByText("-");
    fireEvent.click(decrementButton);
    expect(mockUpdateQuantity).toHaveBeenCalledWith(props.id, 1); // quantidade mockada: 2 - 1
  });

  it("should call removeItem when trash icon is clicked", () => {
    render(<BookItem {...props} />);
    const deleteIcon = screen.getByRole("img", { name: /delete/i });
    fireEvent.click(deleteIcon);
    expect(mockRemoveItem).toHaveBeenCalledWith(props.id);
  });

  it("should not render controls when remove is false", () => {
    render(<BookItem {...props} remove={false} />);
    expect(screen.queryByText("+")).not.toBeInTheDocument();
    expect(screen.queryByText("-")).not.toBeInTheDocument();
    expect(
      screen.queryByRole("img", { name: /delete/i }),
    ).not.toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(<BookItem {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
