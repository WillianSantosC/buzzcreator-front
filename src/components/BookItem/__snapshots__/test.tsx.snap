// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookItem /> matches snapshot 1`] = `
<div
  class="p_token(spacings.small) bd-b_0.1rem_solid_token(colors.lightGray) mdMI:d_flex BookItemWrapper"
>
  <div
    class="d_flex flex-d_column mdMI:flex-d_row ai_center"
  >
    <div
      class="flex-sh_0 mr_1.2rem mdMI:mr_token(spacings.xsmall) w_9.6rem mdMI:w_15rem h_5.6rem mdMI:h_7rem [&_img]:w_100% [&_img]:h_100% [&_img]:obj-f_cover"
    >
      <img
        alt="The Pragmatic Programmer"
        height="100"
        src="/img/books.jpg"
        width="100"
      />
    </div>
    <div
      class="d_flex flex-d_column ai_flex-start"
    >
      <h3
        class="fs_token(font.sizes.small) mdMI:fs_2rem lh_token(font.sizes.small) mdMI:lh_2rem fw_token(font.bold) c_token(colors.black) mb_token(spacings.xxsmall)"
      >
        The Pragmatic Programmer
      </h3>
      <div
        class="d_flex ai_center gap_1rem w_20% jc_space-between"
      >
        <div
          class="c_token(colors.white) p_0.2rem_token(spacings.xxsmall) bg-c_token(colors.secondary) bdr_token(border.radius) fs_token(font.sizes.small) fw_token(font.bold)"
        >
          R$Â 120,00
        </div>
        <div
          class="d_flex ai_center jc_space-between gap_1rem"
        >
          <div
            class="d_flex ai_center jc_space-between gap_1rem bd_solid_0.1rem_token(colors.darkGray) bdr_token(border.radius) p_0.3rem [&_button]:bd_none [&_button]:ring_none [&_button]:p_0.3rem_0.8rem [&_button]:m_0 [&_button]:font_bold [&_button]:fs_token(font.sizes.xsmall) [&_button]:bg_token(colors.lightGray) [&_button]:bdr_token(border.radius)"
          >
            <button>
              -
            </button>
            <span>
              2
            </span>
            <button>
              +
            </button>
          </div>
          <img
            alt="Delete"
            class="w_2rem h_2rem cursor_pointer"
            src="/img/trash-bl.svg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
